psasim: use (UNIX/domain) sockets instead of SystemV RPC for messaging

valeriosetti opened the pull request:
## Description

This PR replaces SystemV RPC messaging with UNIX/domain sockets as base communication system in `psasim`.

I'm opening this PR as draft because there might be better performance from other communication systems (see: https://github.com/goldsborough/ipc-bench.git for comparison of IPC). However since this PR also reshapes the core part of `psasim` I would like to get some feedback from reviewers in the meantime.

## PR checklist

- [ ] **changelog** not required
- [ ] **3.6 backport** not required
- [ ] **2.28 backport** not required
- [ ] **tests** not required

gilles-peskine-arm commented on tests/psa-client-server/psasim/include/client.h:
```diff
@@ -1,5 +1,3 @@
-/* PSA Firmware Framework client header for psasim. */
-
 /*
```
Part of the reason we're using psasim for client-server tests is that it implements the functional part of the [PSA firmware framework](https://developer.arm.com/documentation/den0063/latest/) specification. This means that the glue code to turn crypto function calls into remote procedure calls could, in principle, work on top of other PSA firmware framework implementations, and we could potentially productize it in the future.

This pull request removes or changes many standard macros and functions such as `PSA_FRAMEWORK_VERSION`, `psa_connect`, `psa_call`, etc. Is that really necessary?

gilles-peskine-arm replied:
What I don't understand is how changing the communication _API_ and removing unrelated definitions helps with changing the communication method.

tom-cosgrove-arm replied:
A whole lot of things that weren't necessary were removed, thereby simplifying the code. That made changing the communication method easier. Why would we want to maintain parts of the code that aren't necessary?

gilles-peskine-arm replied:
To be clear, I don't object to simplifying the implementation of the simulator by removing code, so that it would not do anything useful except run our tests. What I object to is changing the _interface_ of the simulator, which means that our glue code is no longer talking to a PSA framework implementation, but to some custom communication implementation. This reduces the confidence we could have that our code would work with a real-world communication interface.

For example, if `psa_connect` only supports a single SID and a single version, I want to keep the prototype `psa_handle_t psa_connect(uint32_t sid, uint32_t version)`. I don't mind if the implemenation is
```
if (version != 0) return 0;
if (sid != 42) return 0;
global_fd = open(...);
if (global_fd < 0) return 0;
return 1;
```

gilles-peskine-arm replied:
As of 8111ab70dea9d544330c533ba8bb06b9e619b368 I'm happy that we're close enough to the official PSA framework API that when we build our client-server code on top of it, it's a good enough test of whether it could work on a real-world IPC implementation.

I'd prefer to have fewer modifications to the header files, so that we don't risk diverging more in the future. But for now we haven't diverged too much in my opinion.

adeaarm replied:
I agree with Gilles that ideally being as close as possible with the PSA FF interfaces in terms of the client-server comms is desirable for end users (with end users here I mean projects that consume this config). So I would as well recommend, if possible, to minimize the alterations made to the headers / prototypes. One possible use case of this is indeed that others project which are more strict with FF compliance will use these test configs to speed up validation without resorting to going on actual HW (or models of the HW, which might not be available upstream), so having less moving parts, e.g. get to compile with different headers from real world, as possible will be beneficial.

As a side note, I understand you're not serializing / deserializing as TF-M does, and this is not a problem, but did you evaluate/compare against how TF-M does it? In that case, if you can share any analysis it would be great for me to understand if we can improve on TF-M side.

valeriosetti replied:
> One possible use case of this is indeed that others project which are more strict with FF compliance will use these test configs to speed up validation without resorting to going on actual HW

If with "configs" here you meant the Mbed TLS ones, then they have no impact on the client/server communication. Setting up the proper configs for Mbed TLS libraries is done in `all.sh` as for all other test components; `psasim` only implements the communication between client and server. So replacing `psasim` with any other implementation should work just fine.

> So I would as well recommend, if possible, to minimize the alterations made to the headers / prototypes

As far as I understood (and seen for example in Zephyr) anyone who provides a server implementation for PSA crypto also provides the "interface" counterpart to be used on the client side. This is done to make the 2 parties to work together both from the communication interface point of view (IPC vs trustzone vs something else) as well as from the serialization/deserialization one.

> As a side note, I understand you're not serializing / deserializing as TF-M does, and this is not a problem, but did you evaluate/compare against how TF-M does it?

Yes, when we started working on this we took a look at how the (de)serialization is done in TF-M in Zephyr, but it seemed to us that in that case files were manually created[^1] whereas we were looking for something automatically generated by a script. @tom-cosgrove-arm took care of this part, so I only roughly know how it's working, but there is no intentional improvement against the TF-M solution. It's just a different implementation where client/server interface functions for PSA are automatically generated starting from "crypto.h".
If we don't like this solution, I think that in the future we can migrate `psasim` to use the same (de)serialization pattern as TF-M, but IMO this is definitely out of scope for this PR.

### Final remark

It seems that with this PR I raised a lot of concerns so I would like to spend a few words on it.
This PR belongs to an EPIC whose goal was to prove that Mbed TLS can be built and tested in a "pure crypto client" scenario (i.e. `CRYPTO_CLIENT && !CRYPTO_C`). Up to 3.6 release there was no official statement for this to be supported, whereas now we can claim that "pure crypto client" can be built and tested like any other configuration used in test components. So I think the goal is (or will be - there are still couple of PRs ongoing) achieved.
`psasim` is where we started to get to this point, but:

- after playing a bit with it in the initial PRs we realized that it was not really tested (we fixed some bugs) so my conclusion was: if it has not been extensively tested from a functional behavior point of view, how can it claim that it's PSA-FF compliant?
- as I said above, in my understanding the goal of the EPIC was to prove that a "pure crypto client" configuration was possible in Mbed TLS, no matter how the communication between the 2 parties was implemented.

That's why I thought that in the end heavily reshaping it could not be too bad.
Should this be the final state then? Not necessarily. If in the future we find/design a real/good PSA-FF implementation, then we can easily replace `psasim` with that one[^2] and run tests in the same way. But IMO this is out of scope for this PR (and perhaps EPIC).


[^1]: of course please apologize if this is not correct, but none of us was an expert in TF-M so we might have overlooked features/behaviors there.
[^2]: replacing both client and server interfaces, of course.

valeriosetti commented:
Following the merge of #8226 and the update of #9427 (on which this PR depends on)  this PR was updated. If the CI is OK then  the PR is ready for reviews.

valeriosetti commented:
Only InternalCI reported a failure, but the OpenCI is fine, so I think this is ready for reviews :)

minosgalanakis reviewed (dismissed):
Looks go to me. The internal CI failures are unrelated

> The internal CI is breaking for network issues:

> ERROR: Could not install packages due to an OSError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))

valeriosetti commented:
Rebased after #9427 was merged, the PR is ready for reviews again :)

minosgalanakis reviewed (approved):
LGTM

tom-cosgrove-arm commented on tests/psa-client-server/psasim/include/service.h:
```diff
@@ -8,246 +6,47 @@
- *                              a polling timeout is used.
- */
-psa_signal_t psa_wait(psa_signal_t signal_mask, uint32_t timeout);
+/* Implemented functions */
```
Again, should we not leave the doc comments in?

tom-cosgrove-arm commented on tests/psa-client-server/psasim/include/service.h:
```diff
@@ -8,246 +6,47 @@
-#ifdef __cplusplus
-}
-#endif
+/* Extra functions*/
```
Documentation?

tom-cosgrove-arm commented:
Also, can you rebase this on top of latest development please? There have been some additional types added there, so if this was automatically rebased we wouldn't have the property that running the perl scripts leaves the working directory unchanged

valeriosetti commented:
I rebased on top of `development` and added the documentation that I previously removed for `service.h`. Actually I did not just restored it, but edited where necessary in order to fit with the real implementation.

valeriosetti commented:
Gentle ping @tom-cosgrove-arm @minosgalanakis :)

minosgalanakis commented on tests/psa-client-server/psasim/include/client.h:
```diff
@@ -8,68 +6,42 @@
-#define PSA_VERSION_NONE       (0)
 
-/* PSA response types */
-#define PSA_CONNECTION_REFUSED PSA_ERROR_CONNECTION_REFUSED
```
I would add a small block comment explaining the [summary and reasoning](https://github.com/Mbed-TLS/mbedtls/pull/9371/files#r1709525447) of the discussion on this issue.

There is definately work that needs to be done if PSA-SIM is picked up by other projects, so anything that could point contributors the right direction is welcome.

valeriosetti replied:
I updated the `README` file in the `psasim` folder with the current status. I thought that would be a better place to talk about `psasim` overall structure.

minosgalanakis reviewed (commented):
Looks good and almost ready for approval. I would only like to request an extra documentation step for the design discussion we had in this issue, and if possible a re-base on the latest base?

valeriosetti commented:
I just rebased to solve a conflict with `development`. The PR is ready for reviews again.

valeriosetti commented on tests/psa-client-server/psasim/src/psa_ff_server.c:
```diff
@@ -133,7 +133,7 @@ size_t psa_write(psa_handle_t msg_handle, uint32_t outvec_idx, const void *buffe
 
     if (atomic_load(&(shared_memory->owner)) != SHARED_MEMORY_OWNER_SERVER) {
         ERROR("Cannot write on shared memory while it's owned by the client");
-        return PSA_ERROR_BAD_STATE;
```
The documentation of this function is now:
```
 * \retval >0                   On success the number of written bytes is returned.
 * \retval 0                    On failure, which can happen if:
 *                              - outvec_idx is invalid, i.e. outvec_idx > PSA_MAX_IOVEC,
 *                              - the server tries to write a message when it's not
 *                                owning the shared memory.
 ```
so I think that returning 0 in case of errors makes sense. Moreover the return type is `size_t`, which is unsigned so returning a signed value might "erroneously" appear as a large positive value for anyone checking the return value. I know that it would be unpractical as value since it would be very large, but still...

minosgalanakis reviewed (approved):
LGTM

valeriosetti commented:
@tom-cosgrove-arm is the documentation update I did since your [last review](https://github.com/Mbed-TLS/mbedtls/pull/9371#issuecomment-2363253999) OK for you? Is there any change you would like to add?

valeriosetti commented:
I'm about to start the review of #9602 which changes some files in `psasim`. It would be nice to merge this PR so that (a) we can avoid conflicts and (b) we can move #9400 forward.
@tom-cosgrove-arm

valeriosetti commented:
I just noticed that this PR has `priority-medium`. Can I move it to `priority-high` so that we can merge it?
Merging is necessary to finally add test parity introduced the by #9400 in the CI
